### PR TITLE
Extract client secret hasher.

### PR DESCRIPTION
**Pull request by SebastianStehle**
In my case the client secrets are already hashed, so I don't need this code. I can just create a custom application manager (and it works) but it is easier like this, I think.

Also makes unit testing easier.

### CODE REVIEW

File: src/OpenIddict.Abstractions/Managers/NullClientSecretHasher.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        /// <inheritdoc />
+        public ValueTask<string> ObfuscateClientSecretAsync(string secret, CancellationToken cancellationToken = default)
+        {
+            return new ValueTask<string>(secret);
+        }
+
+        /// <inheritdoc />
+        public ValueTask<bool> ValidateClientSecretAsync(string secret, string comparand, CancellationToken cancellationToken = default)
+        {
+            var equals = string.Equals(secret, comparand);
```
Comment (kevinchalet) [inline src/OpenIddict.Abstractions/Managers/NullClientSecretHasher.cs:29]: Such equality checks are optimized to return `false` as soon as the first non-matching char is encountered, which makes this implementation prone to statistical/timing attacks. It's definitely not something that should be provided by OpenIddict itself.

**Review by kevinchalet — COMMENTED** [review]
Thanks for your PR 😃 

While it's not clear to me how it helps make things clearer compared to the current approach, I'm actually much more concerned by the `NullClientSecretHasher` you added for two reasons:
  - It doesn't use a time-constant comparison, which makes it vulnerable to timing attacks. The current OpenIddict implementation doesn't suffer from this issue as it uses a time-constant comparison.
  - Comparing the 2 client secret hashes makes your app vulnerable to pass-the-hash attacks (the hash becomes the secret). The OpenIddict implementation doesn't have this issue, as it requires sending the real secret and not a hash (also, it's not a simple hash but the output of a PBKDF, which slows down the secret validation and helps mitigate brute force attacks).

Closing this PR for now.

**Comment by SebastianStehle** [comment]
I can just remove the null-implementation, but it would it make easier to override it. What do you mean with "timing attacks"?

It is just a minor improvements, I know, but I think that the current application manager makes more high-level stuff and should delegate this low-level things to other services anyway. Like password policies and so on the UserManager from ASP.NET Core Identity.

**Comment by kevinchalet** [comment]
> but it would it make easier to override it

Not making it easy to override was actually deliberate: folks are often tempted to customize things and this one is hard to customize "safely", which is why it's only doable by subclassing the manager itself.

> What do you mean with "timing attacks"?

Here's a good overview of what timing attacks are: https://en.wikipedia.org/wiki/Timing_attack

**Comment by SebastianStehle** [comment]
> Not making it easy to override was actually deliberate: folks are often tempted to customize things and this one is hard to customize "safely", which is why it's only doable by subclassing the manager itself.

I see. It is the only way that I found to be compatible with my current identity-server implementation, because the secrets are from another databases (or multiple databases actually). I guess another approach could be to obfuscate the secrets on the fly, when they are loaded from the database. But they I would access to the hasher algorithm as well and I am not sure what it means for performance.

**Pull request closed without merging**